File: native_runtime/runtime_voice_reactive_motion.py
from __future__ import annotations

import math
from typing import Any


VOICE_REACTIVE_MOTION_MODES = {"off", "soft", "cinematic"}
DEFAULT_VOICE_REACTIVE_MOTION_CONFIG = {
    "voiceReactiveMotionMode": "soft",
    "voiceReactiveMotionIntensity": 58,
    "voiceReactiveMotionSensitivity": 62,
}
VOICE_REACTIVE_MOTION_PROFILES = {
    "off": {"scaleBoost": 0.0, "liftPercent": 0.0},
    "soft": {"scaleBoost": 0.006, "liftPercent": 0.24},
    "cinematic": {"scaleBoost": 0.014, "liftPercent": 0.55},
}
MAX_PCM_ANALYSIS_BYTES = 16 * 1024 * 1024


def _clamp(value: float, minimum: float, maximum: float) -> float:
    return max(minimum, min(maximum, value))


def _safe_number(value: Any, fallback: float) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return float(fallback)


def _round_pose_value(value: float) -> float:
    return round(float(value), 3)


def _config_source(value: dict | None) -> dict:
    source = value if isinstance(value, dict) else {}
    nested = source.get("gameUiConfig")
    return nested if isinstance(nested, dict) else source


def _visual_comfort_motion_scale(value: str | None) -> float:
    if value == "static":
        return 0.0
    if value == "gentle":
        return 0.35
    return 1.0


def get_safe_voice_reactive_motion_mode(value: Any) -> str:
    mode = str(value or DEFAULT_VOICE_REACTIVE_MOTION_CONFIG["voiceReactiveMotionMode"]).strip().lower()
    return mode if mode in VOICE_REACTIVE_MOTION_MODES else DEFAULT_VOICE_REACTIVE_MOTION_CONFIG["voiceReactiveMotionMode"]


def sanitize_voice_reactive_motion_config(value: dict | None = None) -> dict:
    source = _config_source(value)
    return {
        "voiceReactiveMotionMode": get_safe_voice_reactive_motion_mode(source.get("voiceReactiveMotionMode")),
        "voiceReactiveMotionIntensity": int(
            round(
                _clamp(
                    _safe_number(
                        source.get("voiceReactiveMotionIntensity"),
                        DEFAULT_VOICE_REACTIVE_MOTION_CONFIG["voiceReactiveMotionIntensity"],
                    ),
                    0,
                    100,
                )
            )
        ),
        "voiceReactiveMotionSensitivity": int(
            round(
                _clamp(
                    _safe_number(
                        source.get("voiceReactiveMotionSensitivity"),
                        DEFAULT_VOICE_REACTIVE_MOTION_CONFIG["voiceReactiveMotionSensitivity"],
                    ),
                    0,
                    100,
                )
            )
        ),
    }


def normalize_voice_reactive_level(raw_level: Any, sensitivity: Any = 62, previous_level: Any = 0) -> float:
    safe_raw_level = _clamp(_safe_number(raw_level, 0), 0, 1)
    safe_sensitivity = _clamp(_safe_number(sensitivity, 62), 0, 100) / 100
    threshold = 0.07 - safe_sensitivity * 0.052
    gain = 3.2 + safe_sensitivity * 4.4
    target = _clamp((safe_raw_level - threshold) * gain, 0, 1)
    previous = _clamp(_safe_number(previous_level, 0), 0, 1)
    smoothing = 0.58 if target >= previous else 0.2
    return _round_pose_value(previous + (target - previous) * smoothing)


def build_native_voice_reactive_motion_pose(
    *,
    character_id: str | None,
    active_character_id: str | None,
    voice_active: bool,
    voice_level: Any,
    game_ui_config: dict | None = None,
    visual_comfort_mode: str = "standard",
    is_leaving: bool = False,
) -> dict:
    config = sanitize_voice_reactive_motion_config(game_ui_config)
    safe_character_id = str(character_id or "").strip()
    safe_active_character_id = str(active_character_id or "").strip()
    motion_scale = _visual_comfort_motion_scale(visual_comfort_mode)
    active = bool(
        config["voiceReactiveMotionMode"] != "off"
        and voice_active
        and not is_leaving
        and safe_character_id
        and safe_character_id == safe_active_character_id
        and motion_scale > 0
    )
    level = _clamp(_safe_number(voice_level, 0), 0, 1) if active else 0.0
    activity = level * (config["voiceReactiveMotionIntensity"] / 100) * motion_scale
    profile = VOICE_REACTIVE_MOTION_PROFILES[config["voiceReactiveMotionMode"]]
    return {
        "mode": config["voiceReactiveMotionMode"],
        "active": active,
        "level": _round_pose_value(level),
        "mouthOpen": _round_pose_value(activity),
        "scaleMultiplier": _round_pose_value(1 + profile["scaleBoost"] * activity),
        "offsetYPercent": _round_pose_value(-profile["liftPercent"] * activity),
    }


def _pcm_sample_view(raw: bytes, mixer_format: int):
    bits = abs(int(mixer_format or 0))
    signed = int(mixer_format or 0) < 0
    format_code = {
        (8, True): "b",
        (8, False): "B",
        (16, True): "h",
        (16, False): "H",
        (32, True): "i",
        (32, False): "I",
    }.get((bits, signed))
    if not format_code:
        return None, 1.0, 0.0
    try:
        samples = memoryview(raw).cast(format_code)
    except (TypeError, ValueError):
        return None, 1.0, 0.0
    maximum = float((1 << (bits - 1)) - 1 if signed else (1 << bits) - 1)
    center = 0.0 if signed else maximum / 2
    normalizer = maximum if signed else max(center, 1.0)
    return samples, normalizer, center


def build_pcm_voice_envelope(
    raw: bytes | bytearray | memoryview | None,
    mixer_init: tuple | list | None,
    *,
    window_ms: int = 40,
    samples_per_window: int = 64,
) -> list[float]:
    if not raw or len(raw) > MAX_PCM_ANALYSIS_BYTES or not isinstance(mixer_init, (tuple, list)) or len(mixer_init) < 3:
        return []
    sample_rate = max(1, int(mixer_init[0] or 0))
    mixer_format = int(mixer_init[1] or 0)
    channels = max(1, int(mixer_init[2] or 1))
    samples, normalizer, center = _pcm_sample_view(bytes(raw), mixer_format)
    if samples is None:
        return []
    frames_per_window = max(1, int(sample_rate * max(10, int(window_ms)) / 1000))
    sample_count_per_window = frames_per_window * channels
    envelope: list[float] = []
    for start in range(0, len(samples), sample_count_per_window):
        end = min(len(samples), start + sample_count_per_window)
        if end <= start:
            break
        stride = max(channels, ((end - start) // max(1, samples_per_window) // channels) * channels)
        energy = 0.0
        count = 0
        for index in range(start, end, stride):
            sample = (float(samples[index]) - center) / max(normalizer, 1.0)
            energy += sample * sample
            count += 1
        envelope.append(_round_pose_value(_clamp(math.sqrt(energy / max(count, 1)), 0, 1)))
    return envelope


def _fallback_voice_level(elapsed_ms: int) -> float:
    time_seconds = max(0, int(elapsed_ms)) / 1000
    pulse = abs(math.sin(time_seconds * 11.7)) * 0.55 + abs(math.sin(time_seconds * 4.3 + 0.8)) * 0.45
    return 0.035 + pulse * 0.095


class NativeVoiceReactiveMotionController:
    """Analyze the current voice clip without coupling audio decoding to the renderer."""

    def __init__(self) -> None:
        self.character_id = ""
        self.started_at_ms = 0
        self.duration_ms = 0
        self.window_ms = 40
        self.envelope: list[float] = []
        self.current_level = 0.0
        self.last_level_at_ms = -1

    def start(self, sound, character_id: str | None, now_ms: int, mixer_init=None) -> None:
        self.character_id = str(character_id or "").strip()
        self.started_at_ms = int(now_ms or 0)
        self.current_level = 0.0
        self.last_level_at_ms = -1
        try:
            self.duration_ms = max(0, round(float(sound.get_length()) * 1000))
        except Exception:
            self.duration_ms = 0
        try:
            raw = sound.get_raw()
        except Exception:
            raw = None
        self.envelope = build_pcm_voice_envelope(raw, mixer_init, window_ms=self.window_ms)

    def stop(self) -> None:
        self.character_id = ""
        self.started_at_ms = 0
        self.duration_ms = 0
        self.envelope = []
        self.current_level = 0.0
        self.last_level_at_ms = -1

    def get_level(self, now_ms: int, voice_active: bool, sensitivity: Any = 62) -> float:
        safe_now_ms = int(now_ms or 0)
        if not voice_active or not self.character_id:
            self.current_level = normalize_voice_reactive_level(0, sensitivity, self.current_level)
            return self.current_level
        if safe_now_ms == self.last_level_at_ms:
            return self.current_level
        elapsed_ms = max(0, safe_now_ms - self.started_at_ms)
        if self.envelope:
            index = min(len(self.envelope) - 1, elapsed_ms // self.window_ms)
            raw_level = self.envelope[index]
        else:
            raw_level = _fallback_voice_level(elapsed_ms)
        self.current_level = normalize_voice_reactive_level(raw_level, sensitivity, self.current_level)
        self.last_level_at_ms = safe_now_ms
        return self.current_level

    def build_render_pose(
        self,
        *,
        character_id: str,
        active_character_id: str | None,
        voice_active: bool,
        game_ui_config: dict | None,
        visual_comfort_mode: str,
        now_ms: int,
        is_leaving: bool = False,
    ) -> dict:
        config = sanitize_voice_reactive_motion_config(game_ui_config)
        level = self.get_level(now_ms, voice_active, config["voiceReactiveMotionSensitivity"])
        return build_native_voice_reactive_motion_pose(
            character_id=character_id,
            active_character_id=active_character_id,
            voice_active=voice_active,
            voice_level=level,
            game_ui_config=config,
            visual_comfort_mode=visual_comfort_mode,
            is_leaving=is_leaving,
        )

File: export_player_template/runtime_voice_reactive_motion.js
// Voice-driven character motion shared by the editor preview and Web Runtime.
// Static sprites receive restrained body motion while model adapters can consume mouthOpen.

export const VOICE_REACTIVE_MOTION_MODE_LABELS = Object.freeze({
  off: "关闭语音演技",
  soft: "自然语音起伏",
  cinematic: "鲜明舞台演技",
});

export const DEFAULT_VOICE_REACTIVE_MOTION_CONFIG = Object.freeze({
  voiceReactiveMotionMode: "soft",
  voiceReactiveMotionIntensity: 58,
  voiceReactiveMotionSensitivity: 62,
});

export const VOICE_REACTIVE_MOTION_PROFILES = Object.freeze({
  off: Object.freeze({ scaleBoost: 0, liftPercent: 0 }),
  soft: Object.freeze({ scaleBoost: 0.006, liftPercent: 0.24 }),
  cinematic: Object.freeze({ scaleBoost: 0.014, liftPercent: 0.55 }),
});

const ACTIVE_CLASS = "is-voice-reactive";
const STYLE_PROPERTIES = Object.freeze([
  "--voice-reactive-level",
  "--voice-reactive-mouth",
  "--voice-reactive-scale",
  "--voice-reactive-lift",
]);

function clamp(value, minimum, maximum) {
  return Math.min(Math.max(value, minimum), maximum);
}

function getSafeNumber(value, fallback) {
  const numeric = Number.parseFloat(value ?? "");
  return Number.isFinite(numeric) ? numeric : fallback;
}

function roundPoseValue(value) {
  return Math.round(Number(value) * 1000) / 1000;
}

function getConfigSource(value) {
  if (value?.gameUiConfig && typeof value.gameUiConfig === "object") {
    return value.gameUiConfig;
  }
  return value && typeof value === "object" ? value : {};
}

function getVisualComfortMotionScale(value) {
  if (value === "static") return 0;
  if (value === "gentle") return 0.35;
  return 1;
}

export function getSafeVoiceReactiveMotionMode(value) {
  const mode = String(value ?? DEFAULT_VOICE_REACTIVE_MOTION_CONFIG.voiceReactiveMotionMode)
    .trim()
    .toLowerCase();
  return Object.hasOwn(VOICE_REACTIVE_MOTION_PROFILES, mode)
    ? mode
    : DEFAULT_VOICE_REACTIVE_MOTION_CONFIG.voiceReactiveMotionMode;
}

export function getVoiceReactiveMotionConfig(value = {}) {
  const source = getConfigSource(value);
  return Object.freeze({
    voiceReactiveMotionMode: getSafeVoiceReactiveMotionMode(source.voiceReactiveMotionMode),
    voiceReactiveMotionIntensity: Math.round(
      clamp(
        getSafeNumber(
          source.voiceReactiveMotionIntensity,
          DEFAULT_VOICE_REACTIVE_MOTION_CONFIG.voiceReactiveMotionIntensity
        ),
        0,
        100
      )
    ),
    voiceReactiveMotionSensitivity: Math.round(
      clamp(
        getSafeNumber(
          source.voiceReactiveMotionSensitivity,
          DEFAULT_VOICE_REACTIVE_MOTION_CONFIG.voiceReactiveMotionSensitivity
        ),
        0,
        100
      )
    ),
  });
}

export function normalizeVoiceReactiveLevel(rawLevel, sensitivity = 62, previousLevel = 0) {
  const safeRawLevel = clamp(getSafeNumber(rawLevel, 0), 0, 1);
  const safeSensitivity = clamp(getSafeNumber(sensitivity, 62), 0, 100) / 100;
  const threshold = 0.07 - safeSensitivity * 0.052;
  const gain = 3.2 + safeSensitivity * 4.4;
  const target = clamp((safeRawLevel - threshold) * gain, 0, 1);
  const previous = clamp(getSafeNumber(previousLevel, 0), 0, 1);
  const smoothing = target >= previous ? 0.58 : 0.2;
  return roundPoseValue(previous + (target - previous) * smoothing);
}

export function buildVoiceReactiveMotionPose({
  characterId,
  activeCharacterId,
  voiceActive = false,
  voiceLevel = 0,
  gameUiConfig = {},
  visualComfortMode = "standard",
  isLeaving = false,
} = {}) {
  const config = getVoiceReactiveMotionConfig(gameUiConfig);
  const safeCharacterId = String(characterId ?? "").trim();
  const safeActiveCharacterId = String(activeCharacterId ?? "").trim();
  const motionScale = getVisualComfortMotionScale(visualComfortMode);
  const active = Boolean(
    config.voiceReactiveMotionMode !== "off" &&
      voiceActive &&
      !isLeaving &&
      safeCharacterId &&
      safeCharacterId === safeActiveCharacterId &&
      motionScale > 0
  );
  const level = active ? clamp(getSafeNumber(voiceLevel, 0), 0, 1) : 0;
  const intensity = config.voiceReactiveMotionIntensity / 100;
  const activity = level * intensity * motionScale;
  const profile = VOICE_REACTIVE_MOTION_PROFILES[config.voiceReactiveMotionMode];

  return Object.freeze({
    mode: config.voiceReactiveMotionMode,
    active,
    level: roundPoseValue(level),
    mouthOpen: roundPoseValue(activity),
    scaleMultiplier: roundPoseValue(1 + profile.scaleBoost * activity),
    offsetYPercent: roundPoseValue(-profile.liftPercent * activity),
  });
}

export function buildVoiceReactiveMotionPresentation(options = {}) {
  const pose = buildVoiceReactiveMotionPose(options);
  return Object.freeze({
    ...pose,
    classNames: Object.freeze(pose.active ? [ACTIVE_CLASS] : []),
    style: [
      `--voice-reactive-level:${pose.level.toFixed(3)}`,
      `--voice-reactive-mouth:${pose.mouthOpen.toFixed(3)}`,
      `--voice-reactive-scale:${pose.scaleMultiplier.toFixed(3)}`,
      `--voice-reactive-lift:${pose.offsetYPercent.toFixed(3)}%`,
    ].join(";"),
  });
}

function getRmsLevel(analyser, data) {
  if (!analyser || !data) return null;
  analyser.getByteTimeDomainData(data);
  let sum = 0;
  for (let index = 0; index < data.length; index += 1) {
    const sample = (data[index] - 128) / 128;
    sum += sample * sample;
  }
  return clamp(Math.sqrt(sum / Math.max(data.length, 1)), 0, 1);
}

function getFallbackVoiceLevel(audio, nowMs) {
  if (!audio || audio.paused || audio.ended) return 0;
  const time = Number.isFinite(audio.currentTime) ? audio.currentTime : nowMs / 1000;
  const pulse = Math.abs(Math.sin(time * 11.7)) * 0.55 + Math.abs(Math.sin(time * 4.3 + 0.8)) * 0.45;
  return 0.035 + pulse * 0.095;
}

function resetMotionNode(node) {
  if (!node) return;
  node.classList?.remove?.(ACTIVE_CLASS);
  STYLE_PROPERTIES.forEach((property) => node.style?.removeProperty?.(property));
  if (node.dataset) {
    delete node.dataset.voiceReactiveLevel;
    delete node.dataset.mouthOpen;
  }
}

export function createVoiceReactiveMotionController(options = {}) {
  const scope = options.scope ?? globalThis;
  const animationApi = options.animationApi ?? scope;
  const resolveRoot = typeof options.resolveRoot === "function"
    ? options.resolveRoot
    : () => options.root ?? scope.document ?? null;
  const audioGraphs = new WeakMap();
  let audioContext = null;
  let activeAudio = null;
  let activeCharacterId = "";
  let gameUiConfig = {};
  let visualComfortMode = "standard";
  let currentLevel = 0;
  let frameHandle = null;
  let boundAudio = null;

  function cancelFrame() {
    if (frameHandle == null) return;
    if (typeof animationApi.cancelAnimationFrame === "function") {
      animationApi.cancelAnimationFrame(frameHandle);
    } else if (typeof animationApi.clearTimeout === "function") {
      animationApi.clearTimeout(frameHandle);
    }
    frameHandle = null;
  }

  function scheduleFrame(callback) {
    if (typeof animationApi.requestAnimationFrame === "function") {
      frameHandle = animationApi.requestAnimationFrame(callback);
      return;
    }
    if (typeof animationApi.setTimeout === "function") {
      frameHandle = animationApi.setTimeout(() => callback(Date.now()), 33);
    }
  }

  function connectAudioGraph(graph) {
    if (!graph || graph.connected) return graph;
    try {
      graph.sourceNode.connect(graph.analyser);
      graph.analyser.connect(audioContext.destination);
      graph.connected = true;
      return graph;
    } catch (_error) {
      try {
        graph.sourceNode.disconnect();
        graph.analyser.disconnect();
      } catch (_disconnectError) {
        // A partially connected graph is still safe to abandon.
      }
      return null;
    }
  }

  function disconnectAudioGraph(audio) {
    const graph = audio ? audioGraphs.get(audio) : null;
    if (!graph?.connected) return;
    try {
      graph.sourceNode.disconnect();
    } catch (_error) {
      // Already disconnected by the browser.
    }
    try {
      graph.analyser.disconnect();
    } catch (_error) {
      // Already disconnected by the browser.
    }
    graph.connected = false;
  }

  function getAudioGraph(audio) {
    if (!audio) return null;
    const cached = audioGraphs.get(audio);
    if (cached) return connectAudioGraph(cached);
    try {
      if (!audioContext) {
        const AudioContextConstructor = options.AudioContext ?? scope.AudioContext ?? scope.webkitAudioContext;
        audioContext = typeof options.audioContextFactory === "function"
          ? options.audioContextFactory()
          : AudioContextConstructor
            ? new AudioContextConstructor()
            : null;
      }
      if (!audioContext?.createMediaElementSource || !audioContext?.createAnalyser) return null;
      const sourceNode = audioContext.createMediaElementSource(audio);
      const analyser = audioContext.createAnalyser();
      analyser.fftSize = 256;
      analyser.smoothingTimeConstant = 0.42;
      const graph = {
        sourceNode,
        analyser,
        data: new Uint8Array(analyser.fftSize),
        connected: false,
      };
      audioGraphs.set(audio, graph);
      return connectAudioGraph(graph);
    } catch (_error) {
      return null;
    }
  }

  function resetAllNodes() {
    const root = resolveRoot();
    root?.querySelectorAll?.(`.sprite-card.${ACTIVE_CLASS}`)?.forEach?.(resetMotionNode);
  }

  function applyCurrentPose() {
    const root = resolveRoot();
    if (!root?.querySelectorAll) return;
    const config = getVoiceReactiveMotionConfig(gameUiConfig);
    root.querySelectorAll(".sprite-card[data-character-id]").forEach((node) => {
      const characterId = String(node.dataset?.characterId ?? "");
      const presentation = buildVoiceReactiveMotionPresentation({
        characterId,
        activeCharacterId,
        voiceActive: Boolean(activeAudio && !activeAudio.paused && !activeAudio.ended),
        voiceLevel: currentLevel,
        gameUiConfig: config,
        visualComfortMode,
        isLeaving: node.classList?.contains?.("is-leaving"),
      });
      if (!presentation.active) {
        resetMotionNode(node);
        return;
      }
      node.classList.add(...presentation.classNames);
      presentation.style.split(";").filter(Boolean).forEach((declaration) => {
        const separator = declaration.indexOf(":");
        node.style.setProperty(declaration.slice(0, separator), declaration.slice(separator + 1));
      });
      node.dataset.voiceReactiveLevel = presentation.level.toFixed(3);
      node.dataset.mouthOpen = presentation.mouthOpen.toFixed(3);
    });
  }

  function finishAudio(audio) {
    if (!audio || audio !== activeAudio) return;
    cancelFrame();
    if (boundAudio === audio) unbindAudioEvents();
    disconnectAudioGraph(audio);
    activeAudio = null;
    activeCharacterId = "";
    currentLevel = 0;
    resetAllNodes();
  }

  function handleAudioPlaying() {
    if (!activeAudio) return;
    getAudioGraph(activeAudio);
    cancelFrame();
    scheduleFrame(tick);
  }

  function handleAudioPause() {
    if (!activeAudio || activeAudio.ended) return;
    cancelFrame();
    currentLevel = 0;
    resetAllNodes();
  }

  function handleAudioEnded() {
    finishAudio(boundAudio);
  }

  function unbindAudioEvents() {
    if (!boundAudio?.removeEventListener) {
      boundAudio = null;
      return;
    }
    boundAudio.removeEventListener("play", handleAudioPlaying);
    boundAudio.removeEventListener("playing", handleAudioPlaying);
    boundAudio.removeEventListener("pause", handleAudioPause);
    boundAudio.removeEventListener("ended", handleAudioEnded);
    boundAudio = null;
  }

  function bindAudioEvents(audio) {
    if (audio === boundAudio) return;
    unbindAudioEvents();
    boundAudio = audio ?? null;
    if (!boundAudio?.addEventListener) return;
    boundAudio.addEventListener("play", handleAudioPlaying);
    boundAudio.addEventListener("playing", handleAudioPlaying);
    boundAudio.addEventListener("pause", handleAudioPause);
    boundAudio.addEventListener("ended", handleAudioEnded);
  }

  function tick(timestamp) {
    frameHandle = null;
    if (!activeAudio) {
      resetAllNodes();
      return;
    }
    if (activeAudio.ended) {
      finishAudio(activeAudio);
      return;
    }
    if (activeAudio.paused) {
      currentLevel = 0;
      resetAllNodes();
      return;
    }
    const config = getVoiceReactiveMotionConfig(gameUiConfig);
    const graph = getAudioGraph(activeAudio);
    const rawLevel = getRmsLevel(graph?.analyser, graph?.data) ?? getFallbackVoiceLevel(activeAudio, timestamp);
    currentLevel = normalizeVoiceReactiveLevel(
      rawLevel,
      config.voiceReactiveMotionSensitivity,
      currentLevel
    );
    applyCurrentPose();
    scheduleFrame(tick);
  }

  function start({ audio, characterId, gameUiConfig: nextConfig = {}, visualComfortMode: nextComfort = "standard" } = {}) {
    const config = getVoiceReactiveMotionConfig(nextConfig);
    const nextCharacterId = String(characterId ?? "").trim();
    if (audio !== activeAudio) {
      disconnectAudioGraph(activeAudio);
      currentLevel = 0;
    }
    activeAudio = audio ?? null;
    activeCharacterId = nextCharacterId;
    gameUiConfig = config;
    visualComfortMode = nextComfort;
    cancelFrame();
    if (!activeAudio || !activeCharacterId || config.voiceReactiveMotionMode === "off" || nextComfort === "static") {
      unbindAudioEvents();
      disconnectAudioGraph(activeAudio);
      activeAudio = null;
      activeCharacterId = "";
      resetAllNodes();
      return false;
    }
    bindAudioEvents(activeAudio);
    getAudioGraph(activeAudio);
    audioContext?.resume?.().catch?.(() => {});
    scheduleFrame(tick);
    return true;
  }

  function stop() {
    cancelFrame();
    unbindAudioEvents();
    disconnectAudioGraph(activeAudio);
    activeAudio = null;
    activeCharacterId = "";
    currentLevel = 0;
    resetAllNodes();
  }

  return Object.freeze({ start, stop, applyCurrentPose });
}

const runtimeVoiceReactiveMotionApi = Object.freeze({
  VOICE_REACTIVE_MOTION_MODE_LABELS,
  DEFAULT_VOICE_REACTIVE_MOTION_CONFIG,
  VOICE_REACTIVE_MOTION_PROFILES,
  getSafeVoiceReactiveMotionMode,
  getVoiceReactiveMotionConfig,
  normalizeVoiceReactiveLevel,
  buildVoiceReactiveMotionPose,
  buildVoiceReactiveMotionPresentation,
  createVoiceReactiveMotionController,
});

globalThis.CanvasiaRuntimeVoiceReactiveMotion = runtimeVoiceReactiveMotionApi;
